packages/connect/src/core/index.ts

```diff
@@ -1149,3 +1162,341 @@ const disableWebUSBTransport = async () => {
+    if (support.reboot_and_upgrade) {
+        flow = 'reboot_and_upgrade';
+    } else if (support.reboot_and_wait) {
+        flow = 'reboot_and_wait';
+    }
+
+    log.debug('onCallFirmwareUpdate flow selected: ', flow);
+
+    await device.acquire();
+    let binary: ArrayBuffer | undefined;
+    // parametrized getBinary to save some lines of code
+    const getBinaryHelper = async (forDevice: Device) => {
+        if (!forDevice.firmwareRelease) {
+            throw TypedError('Runtime', 'device.firmwareRelease is not set');
+        }
+        const btcOnly = params.btcOnly || device.firmwareType === 'bitcoin-only';
+
+        log.debug(
+            'onCallFirmwareUpdate loading binary',
+            'intermediaryVersion',
+            forDevice.firmwareRelease.intermediaryVersion,
+            'version',
+            forDevice.firmwareRelease.release.version,
+            'btcOnly',
+            btcOnly,
+        );
+
+        postMessage(
+            createUiMessage(UI.FIRMWARE_PROGRESS, {
+                device: device.toMessageObject(),
+                operation: 'downloading',
+                progress: 0,
+            }),
+        );
+        return getBinary({
+            // features and releases are used for sanity checking
+            features: forDevice.features,
+            releases: getReleases(forDevice.features?.internal_model),
+            // version argument is used to find and fetch concrete release from releases list,
+            // at the moment, version is not passed in params, so which version is going to be installed is solely
+            // decided by connect internals
+            version: forDevice.firmwareRelease.release.version,
+            btcOnly,
+            baseUrl: params.baseUrl!,
+            intermediaryVersion: forDevice.firmwareRelease.intermediaryVersion,
+        }).then(res => {
+            postMessage(
+                createUiMessage(UI.FIRMWARE_PROGRESS, {
+                    device: device.toMessageObject(),
+                    operation: 'downloading',
+                    progress: 100,
+                }),
+            );
+            return res;
+        });
+    };
+
+    binary = params.binary || (await getBinaryHelper(device));
+
+    log.debug('onCallFirmwareUpdate initial binary downloaded', flow);
+
+    if (flow !== 'manual') {
+        const rebootToBooloderParams: PROTO.RebootToBootloader = {};
+        let languageBlob: ArrayBuffer | undefined;
+        // boot command REBOOT_AND_UPGRADE (1) can only be used when upgrading to higher version - so no reinstall
+        if (flow === 'reboot_and_upgrade') {
+            // download firmware
+
+            rebootToBooloderParams['boot_command'] = 1;
+            rebootToBooloderParams['firmware_header'] = Buffer.from(binary.slice(0, 6000)).toString(
+                'hex',
+            );
+        }
+        if (support.language_data_length) {
+            languageBlob = await getLanguage({
```

Comment:
If I read this correctly, `languageBlob` is fetched and transferred into device everytime when supported in non-manual flow. Is it ok even with custom firmware (`params.binary`) passed? Because then `device.release.version` is used for the language but it's not the final fw version as fw installation cycle ends after first iteration in that case.

Reply:
if params.binary is passed, language update branch should be imho skipped